### DISCUSSION TITLE
Adding additional objective: Makespan

### DISCUSSION THREAD

**Discussion by bmorlo** [PyVRP/PyVRP#582] [Q&A] [answered]
Hi all!

I'm currently looking at additional makespan minimization for my work i.e., making sure the routes of the best solution the solver proposes are as similar in the amount of clients visited as possible.

I'm already using the lowest possible capacity for the vehicles by which the problem, given the fixed amount of vehicles I have, remains feasible. This somewhat evens out the routes. But not enough.

My approach now would be to determine the maximum difference between the capacity and the number of clients visited across the proposed routes of any solution. This value would then feed into the objective function and should be minimized.

I already looked at the way you guys implemented the price collecting VRP.

I would have two questions:

1) How do you balance the uncollected price of a client against the distance that you saved by not visiting it? Wouldn't the total distance of a solution and the summed uncollected prices be on very different scales? Stated differently: How do you determine the foregone price of a client in relation to the distances to make sure the solver actually considers visiting it? If the foregone price of a client is very small compared to the distance that it takes to visit it, I assume the solver would simply always skip visiting this client. Similarly, if the foregone price of a client is very large compared to the distance that comes with it,  I assume the solver would always choose to visit this client.

2) Could you point me towards the parts of your code that you think I would have to make adjustments to? I feel a bit anxious that adding this feature would be somewhat outside of my abilities. I did check out your "Supporting new VRP variants" section already.

Many thanks & cheers once more to your awesome work!

Benedikt

**Comment by leonlan** [comment]
Hi @bmorlo,

I have a few questions for you:
- Why does the capacity-based approach not work for balancing out the number of clients? Why is it "not enough" and what goes wrong?
- How does makespan relate to the number of clients served per vehicle? Makespan refers to the latest completion time of the "system", i.e., the time that the last vehicle returns to the depot. How does that relate to what you say about "determine the maximum difference between the capacity and the number of clients visited across the proposed routes of any solution?"
- In prize-collecting VRPs, you can choose to not visit all clients. Do you want to allow that?
- More importantly: What are the main goals for your problem? Is it distance minimization? Duration minimization? Is it serving as many clients as possible? Balancing the number of clients per vehicle? Or making sure that all vehicles end roughly at the same time?

I think your approach with a prize-collecting model might be cumbersome and is not really what you want.
But without knowing the exact (business) requirements of your problem it's hard to answer your questions.

**Reply by bmorlo** [reply]
Hi @leonlan,

thanks for your detailed comments! Apologies for not providing you with more details. Please find below my italic answers to your questions.

`Why does the capacity-based approach not work for balancing out the number of clients? Why is it "not enough" and what goes wrong?`

_>> I'm assuming that all clients have delivery = 1. If I then have e.g. 12 clients to be served by 5 vehicles, I would set the capacity of each vehicle to ceiling(12/5) = 3. But 3 * 5 = 15, which leaves the solver with the possibility to only work with 4 vehicles (since 3 * 4 = 12 already). This usually results in 4 tours with 3 clients per tour - which is of course optimal. However, I would like to have all five vehicles utilized by e.g. 3 tours with 2 clients and 2 tours with 3 clients._

`How does makespan relate to the number of clients served per vehicle? Makespan refers to the latest completion time of the "system", i.e., the time that the last vehicle returns to the depot. How does that relate to what you say about "determine the maximum difference between the capacity and the number of clients visited across the proposed routes of any solution?"`

_>> In the problem that I'm working on, I'm assuming that reaching any client location from anywhere takes less than or equal to one week. Fully independent of the actual distance needed to get to that client, that is. Hence, the number of locations per tour or vehicle directly corresponds to how fast a vehicle returns to the depot._

`In prize-collecting VRPs, you can choose to not visit all clients. Do you want to allow that?`

_>> I was also thinking about just using your prize collecting implementation. This way - please correct me if I'm wrong - I could simply set the 'required' argument of all the clients to False and also put very high foregone prizes. Then I could set the same capacity for all vehicles and (again, with delivery = 1 per client) the solver would give me tours that encompass exactly the amount of clients equalling the capacity. However, in my case, the solver's solution should provide me with tours that "look into the future" as much as possible. Thus, I would really have to visit all clients in my case._

`More importantly: What are the main goals for your problem? Is it distance minimization? Duration minimization? Is it serving as many clients as possible? Balancing the number of clients per vehicle? Or making sure that all vehicles end roughly at the same time?`

_>> It is distance minimization but with the addition of balancing the number of clients per vehicle, which, in my special case, would result in similar ending times of the tours._

I hope this clarifies what I'm trying to achieve. Thanks again.

Best,
Benedikt

**Reply by N-Wouda** [reply]
> >> I'm assuming that all clients have delivery = 1. If I then have e.g. 12 clients to be served by 5 vehicles, I would set the capacity of each vehicle to ceiling(12/5) = 3. But 3 * 5 = 15, which leaves the solver with the possibility to only work with 4 vehicles (since 3 * 4 = 12 already). This usually results in 4 tours with 3 clients per tour - which is of course optimal. However, I would like to have all five vehicles utilized by e.g. 3 tours with 2 clients and 2 tours with 3 clients.

What's stopping you from defining three vehicles with capacity `2` and two vehicles with capacity `3`? If the vehicles are otherwise identical that'd solve this problem.

**Reply by bmorlo** [reply]
Doing this would imply that I already know which vehicles to equip with a capacity of two and which with a capacity of three for the best solution. Since I would like the solver to decide on that, I cannot do this. Correct me if I'm wrong. :)

**Reply by leonlan** [reply, marked as answer]
The solver can only decide on the things that it has implemented, which does not include any “same-sized route” constraints. You can deal with this constraint by building it into PyVRP yourself or you need to think of some heuristic that help do the load balancing (for example, setting custom capacities).


Without a mathematical formulation of your problem it’s hard to say what exactly is “optimal” in your problem. But you can first use PyVRP to solve the problem. Based on the solution, you can create a heuristic that sets capacity limits and then solve it again.

**Reply by bmorlo** [reply]
You're right. Using a "self-made" capacity heuristic + reoptimization could indeed be a work-around if I cannot get the secondary objective to work.

I'll try both. Thanks & cheers as always!

By the way: How exactly am I supposed to credit your help in my Thesis? For sure I'll cite your paper, but is there anything else I should do? I'm literally building on the ideas that I got from you here...

**Reply by leonlan** [reply]
Citing the paper is sufficient for us! You can also add a few words in your "Acknowledgements" section, which is what I do if discussions with colleagues have helped me to come up with an idea. Also, spreading the word about PyVRP is always appreciated :-).

**Reply by bmorlo** [reply]
Alright! Will do!

**Comment by N-Wouda** [comment]
>How do you balance the uncollected price of a client against the distance that you saved by not visiting it?

To make the distance worth travelling the prizes should be high enough. It really depends on your application what "high enough" entails exactly. In a problem I'm currently working on, my prizes relate to a probability estimate of some event happening at each client location. The estimate is in [0, 1]. I scale that estimate by a parameter $\rho > 0$ to balance it with the distance estimates. Finding the right value of $\rho$ requires some tuning on the topology of the problem.

There are no shortcuts here. It all depends on the specifics of your problem.

>Could you point me towards the parts of your code that you think I would have to make adjustments to? I feel a bit anxious that adding this feature would be somewhat outside of my abilities. I did check out your "Supporting new VRP variants" section already.

Probably the `CostEvaluator`, since that's in charge of evaluating (changes in) the objective value.
